src/main/java/io/jenkins/pluginhealth/scoring/probes/ProbeEngine.java

```diff
@@ -59,33 +58,33 @@ public ProbeEngine(List<Probe> probes, PluginService pluginService) {
         LOGGER.info("Start running probes on all plugins");
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug("Registered probes: {}", probes.stream().map(Probe::key).collect(Collectors.joining(", ")));
-        }
         pluginService.streamAll().parallel()
-            .peek(plugin -> probes.stream()
-                .map(probe -> {
+            .peek(plugin -> {
+                probes.forEach(probe -> {
                     try {
-                        if (probe.requiresRelease() && plugin.getDetails().containsKey(probe.key())) {
-                            ProbeResult probeResult = plugin.getDetails().get(probe.key());
-                            if (probeResult.timestamp() != null && plugin.getReleaseTimestamp().isBefore(probeResult.timestamp())) {
-                                if (LOGGER.isDebugEnabled()) {
-                                    LOGGER.debug("{} requires a release of {} to process it again.", probe.key(), plugin.getName());
-                                }
-                                return probeResult;
+                        final ProbeResult previousResult = plugin.getDetails().get(probe.key());
+                        if (previousResult == null ||
+                            (probe.requiresRelease()
+                                && previousResult.timestamp() != null
+                                && previousResult.timestamp().isBefore(plugin.getReleaseTimestamp()))
+                        ) {
+                            if (LOGGER.isTraceEnabled()) {
+                                LOGGER.trace("Running {} on {}", probe.key(), plugin.getName());
+                            }
+                            final ProbeResult result = probe.apply(plugin);
+                            if (result.status() != ResultStatus.ERROR) {
+                                plugin.addDetails(result);
+                            }
+                        } else {
+                            if (LOGGER.isDebugEnabled()) {
+                                LOGGER.debug("{} requires a release of {} to process it again.", probe.key(), plugin.getName());
                             }
                         }
-                        if (LOGGER.isTraceEnabled()) {
-                            LOGGER.trace("Running {} on {}", probe.key(), plugin.getName());
-                        }
-                        return probe.apply(plugin);
                     } catch (Throwable t) {
                         LOGGER.error("Couldn't run {} on {}", probe.key(), plugin.getName(), t);
-                        return ProbeResult.error(probe.key(), "Could not run probe " + probe.key());
                     }
-                })
-                .filter(result -> result.status() != ResultStatus.ERROR)
-                .forEach(plugin::addDetails))
+                });
+            })
             .forEach(pluginService::saveOrUpdate);
         LOGGER.info("Probe engine has finished");
     }
```

Comment:
I need to add more tests for this file, more specifically, for the logic inside this `run()` method.
I don't see any other use case to cover, how do you decide if there are any more unit tests that can be added?

For example, here's what I know we should write a test for:
1. `final ProbeResult previousResult = plugin.getDetails().get(probe.key());`
     We should test whether the previousResult of a probe is being accessible its next run, you've already written a test for it.
2. The `if` condition on line 66:
    * A test for `null` previousResult
    * A test for non-null previousResult
    * A test for a probe that requires a release, and another which doesn't
    * A test for a probe whose timestamp is before the relaseTimestamp of the plugin, and vice versa
3. `if` condition on line 75, which makes sure we don't call `addDetails` on a probe with `ResultStatus == ERROR`

And for all the above scenarios we have already written tests. Can you please suggest how can we further increase the test coverage?